pkg/controller/addon/addon_controller.go

```diff
@@ -0,0 +1,358 @@
+package addon
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	mf "github.com/jcrossley3/manifestival"
+	"github.com/prometheus/common/log"
+	op "github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	"github.com/tektoncd/operator/pkg/controller/setup"
+	"golang.org/x/xerrors"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	ctrlLog                   = logf.Log.WithName("ctrl").WithName("addon")
+	errPipelineNotReady       = xerrors.Errorf("tekton-pipelines not ready")
+	errAddonVersionUnresolved = xerrors.Errorf("tekton-pipelines not ready")
+)
+
+// Add creates a new Addon Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileAddon{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("addon-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Addon
+	err = c.Watch(&source.Kind{Type: &op.Addon{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watches for secondary resources
+	// currently watching only deployments
+	err = c.Watch(
+		&source.Kind{Type: &appsv1.Deployment{}},
+		&handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &op.Addon{},
+		})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileAddon implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileAddon{}
+
+// ReconcileAddon reconciles a Addon object
+type ReconcileAddon struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	addons map[string]mf.Manifest
+}
+
+// Reconcile reads that state of the cluster for a Addon object and makes changes based on the state read
+// and what is in the Addon.Spec
+func (r *ReconcileAddon) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := requestLogger(req, "addon reconcile")
```

Comment:
nit: reconcile should do .. as the logger already has `ctrl.addon`

Reply:
OK